### PR TITLE
Make direct I/O write use incremental buffer

### PR DESCRIPTION
**Pull request by siying**
Summary: Currently for direct I/O, the large maximum buffer is always allocated. This will be wasteful if users flush the data in much smaller chunks. This diff fix this by changing the behavior of incremental buffer works. When we enlarge buffer, we try to copy the existing data in the buffer to the enlarged buffer, rather than flush the buffer first. This can make sure that no extra I/O is introduced because of buffer enlargement.

Test Plan:Run existing tests.

### CODE REVIEW

File: util/aligned_buffer.h [hunk 1]
```diff
@@ -97,20 +97,25 @@ class AlignedBuffer {
...
+        ~static_cast<uintptr_t>(alignment_ - 1));
 
-    char* p = buf_.get();
-    bufstart_ = reinterpret_cast<char*>(
-      (reinterpret_cast<uintptr_t>(p)+(alignment_ - 1)) &
-      ~static_cast<uintptr_t>(alignment_ - 1));
-    capacity_ = size;
-    cursize_ = 0;
+    if (copy_data) {
+      memcpy(new_bufstart, bufstart_, cursize_);
```
Comment (yuslepukhin) [inline util/aligned_buffer.h]: Where do we make sure that the newCapacity is at least cursize_ for copying purposes?

Comment (siying) [inline util/aligned_buffer.h]: In file util/file_reader_writer.cc, line 128 to 137.

Comment (siying) [inline util/aligned_buffer.h]: But I'll add an assert too.

File: util/file_reader_writer.cc [hunk 2]
```diff
@@ -124,6 +124,17 @@ Status WritableFileWriter::Append(const Slice& data) {
     writable_file_->PrepareWrite(static_cast<size_t>(GetFileSize()), left);
   }
 
+  // See whether we need to enlarge the buffer to avoid the flush
+  if (buf_.Capacity() - buf_.CurrentSize() < left) {
+    for (size_t desired_capacity = buf_.Capacity();
+         desired_capacity < max_buffer_size_; desired_capacity *= 2) {
+      if (desired_capacity - buf_.CurrentSize() >= left) {
```
Comment (lightmark) [inline util/file_reader_writer.cc]: I think we should cut desired_capacity to max_buffer_size_ if desired_capacity > max_buffer_size_ before this line.

Comment (lightmark) [inline util/file_reader_writer.cc]: or `for (***;disired_capacity <= (max_buffer_size >> 1);***) {}`

Comment (siying) [inline util/file_reader_writer.cc]: Good catch. I'll fix it.

**Comment by siying** [comment]
Add a randomized test.

**Pull request closed without merging**